Fix udunits2

michaelkuhn opened the pull request:
The tarball for 2.25 seems to have been updated (see https://github.com/Unidata/UDUNITS-2/commits/master).

adamjstewart commented:
@citibeth I don't know if you want to try to validate the safety of this tarball after the checksum change? You seem to be the only one who comments on these type of changes.

citibeth commented:
@adamjstewart Yes... changes to checksums without a corresponding change inversion number absolutely need to be investigated.  Otherwise, the checksum loses almost all value as a security mechanism.

I looked at the UDUNITS2 website, and it looks very strange.  There are two new commits.  The first one changes the git hash, but doesn't change any files.  The second changes only a date on a changelog.  I would email Steve Emmerson and Ward Fisher at UNIDATA and ask what's going on.  (I suspect this is some kind of automated conversion from an SVN repo they have somewhere).

https://staff.ucar.edu/browse/orgs/UNIDATA

michaelkuhn commented:
I still had the old tarball lying around:

```
$ md5sum 20170328/var/spack/cache/udunits2/udunits2-2.2.25.tar.gz 20170412/var/spack/cache/udunits2/udunits2-2.2.25.tar.gz 
373106a0fcd20c40fc53a975c9fa4fca  20170328/var/spack/cache/udunits2/udunits2-2.2.25.tar.gz
97e51e12d162ca1c7b0ffa339f8395a9  20170412/var/spack/cache/udunits2/udunits2-2.2.25.tar.gz
$ mkdir 373106a0fcd20c40fc53a975c9fa4fca 97e51e12d162ca1c7b0ffa339f8395a9
$ tar -x -C 373106a0fcd20c40fc53a975c9fa4fca -f 20170328/var/spack/cache/udunits2/udunits2-2.2.25.tar.gz
$ tar -x -C 97e51e12d162ca1c7b0ffa339f8395a9 -f 20170412/var/spack/cache/udunits2/udunits2-2.2.25.tar.gz
$ diff -ruN 373106a0fcd20c40fc53a975c9fa4fca 97e51e12d162ca1c7b0ffa339f8395a9
diff -ruN 373106a0fcd20c40fc53a975c9fa4fca/UDUNITS-2-2.2.25/CHANGE_LOG 97e51e12d162ca1c7b0ffa339f8395a9/UDUNITS-2-2.2.25/CHANGE_LOG
--- 373106a0fcd20c40fc53a975c9fa4fca/UDUNITS-2-2.2.25/CHANGE_LOG	2017-01-25 18:59:53.000000000 +0100
+++ 97e51e12d162ca1c7b0ffa339f8395a9/UDUNITS-2-2.2.25/CHANGE_LOG	2017-04-07 22:17:09.000000000 +0200
@@ -1,4 +1,4 @@
-2.2.25	2017-01-25T10:59:53-0700
+2.2.25	2017-04-07T14:17:09-0600
     Added "stamp-vti" and "version.texi" to ".gitignore".
 
 2.2.24	2017-01-24T14:42:49-0700
```

michaelkuhn commented:
By the way: There is no 2.25 at ftp://ftp.unidata.ucar.edu/pub/udunits, which is the official download location. Seems 2.25 is an unreleased snapshot that gets updated regularly. Maybe we should simply remove 2.25.

citibeth commented:
I'd be OK with that.  But again... I'd like to hear from the upstream folks
about what they are doing.

On Wed, Apr 12, 2017 at 1:54 PM, Michael Kuhn <notifications@github.com>
wrote:

> By the way: There is no 2.25 at ftp://ftp.unidata.ucar.edu/pub/udunits,
> which is the official download location. Seems 2.25 is an unreleased
> snapshot that gets updated regularly. Maybe we should simply remove 2.25.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-293657630>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd1cfWyiAKVQAm4A-SSf8ULp-uzggks5rvQ_FgaJpZM4M7uHs>
> .
>

tgamblin commented:
@semmerson @WardF: We're trying to decide the best strategy to package `udunits` 2.2.25 in Spack.  Do you know why the release tarball keeps changing?

semmerson commented:
@tgamblin I afraid I don't understand what you mean by "the release tarball keeps changing". A given version of the UDUNITS package is cast in stone and can't change. If you're referring to SPACK tarball releases, then I'm afraid I can't help you.

Please elucidate.

citibeth commented:
Steven,

The tarball accessed at
https://github.com/Unidata/UDUNITS-2/archive/v2.2.23.tar.gz has changed
recently.  Inspection of the GitHub log of UDUNITS2 shows multiple commits
labelled "2.2.25."  One of the recent git commits changed nothing.  Another
changed the CHANGELOG file.  I'm wondering if the git repo is a mirror of
an SVN repo somewhere else?

I understand that these changes are not of fundamental substance.  But
systems like Spack rely on tarball checksums to ensure the security of our
systems, as we download and install software from hundreds of different
websites.  That is why we noticed the checksum change and are inquiring
about it.

More details on this issue can be found here:

https://github.com/LLNL/spack/pull/3809

Thank you,
-- Elizabeth


On Fri, Apr 14, 2017 at 12:02 PM, Steven Emmerson <notifications@github.com>
wrote:

> @tgamblin <https://github.com/tgamblin> I afraid I don't understand what
> you mean by "the release tarball keeps changing". A given version of the
> UDUNITS package is cast in stone and can't change. If you're referring to
> SPACK tarball releases, then I'm afraid I can't help you.
>
> Please elucidate.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-294180458>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd903N56GN4d8o0JCIxWIZAhPYGPXks5rv5ingaJpZM4M7uHs>
> .
>

semmerson commented:
@citibeth Well, that shouldn't happen. The only reason for it that I can think of is that I didn't follow my own release protocol. I'll strive to ensure that doesn't happen again.

adamjstewart commented:
It sounds like we have confirmation from the developers that this new tarball checksum is safe to use. @citibeth are you okay with me merging this PR?

semmerson commented:
@citibeth I think I see the problem. The GitHub tarball _isn't_ the official release: official releases are at <[ftp://ftp.unidata.ucar.edu/pub/udunits](ftp://ftp.unidata.ucar.edu/pub/udunits)>. Multiple UDUNITS commits on GitHub can have the same version identifier, which is only cast in stone when the official release is made.

adamjstewart commented:
It looks like 2.25 isn't available at that download location. I believe we used to use that download location, but every time a new release came out the old one was removed, so it wasn't reliable. If you have control of this download location, it would be helpful if it contained every release instead of just the latest.

citibeth commented:
Steve,  can you please point us to the official release?  We will ctecksum
that.
On Apr 14, 2017 12:27 PM, "Steven Emmerson" <notifications@github.com>
wrote:

> @citibeth <https://github.com/citibeth> I think I see the problem. The
> GitHub tarball *isn't* the official release: official releases are at <
> ftp://ftp.unidata.ucar.edu/pub/udunits>. Multiple UDUNITS commits on
> GitHub can have the same version identifier, which is only cast in stone
> when the official release is made.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-294185224>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd2QM_n02cr3cgdGH6cDEildXuQM0ks5rv55igaJpZM4M7uHs>
> .
>

semmerson commented:
@adamjstewart UDUNITS 2.2.25 hasn't yet been officially released. You jumped the gun by downloading a commit with that label from GitHub.

citibeth commented:
Thank you.  Then I recommend we remove it from Spack until it is official.
Thanks Steven!
On Apr 14, 2017 2:33 PM, "Steven Emmerson" <notifications@github.com> wrote:

> @adamjstewart <https://github.com/adamjstewart> UDUNITS 2.2.25 hasn't yet
> been officially released. You jumped the gun by downloading a commit with
> that label from GitHub.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-294209792>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd-tFkci_vvNb5fuh6mLprmvqCYngks5rv7v1gaJpZM4M7uHs>
> .
>

semmerson commented:
@adamjstewart The rationale for deleting older official releases is to encourage uses to adopt the latest "bugfix" release (which is denoted in the 3rd field of the version identifier). That rationale seems valid to me and I'd rather not ignore it.

Can you not copy and save any release you value?

adamjstewart commented:
Spack downloads the release and caches it locally, but if I want to download an older release (for whatever reason) and I've never downloaded it before, we won't be able to find it unless we download from GitHub.

tgamblin commented:
@semmerson: I think the rationale is valid, and encouraging users to use the latest patch version is good. For Spack we also want to provide the ability to reproduce an old install, and that *may* require downloading an old patch release.  We're not out to change your policy, though.  There are plenty of projects that delete old releases, and we can't mandate what they do.  

Question: Is a github label set in stone once its corresponding release appears on your FTP site?

@adamjstewart: If @semmerson's answer is yes, we should be able to rely on github labels as long as the releases we list are official.

citibeth commented:
Steven,


> @adamjstewart <https://github.com/adamjstewart> The rationale for
> deleting older official releases is to encourage uses to adopt the latest
> "bugfix" release (which is denoted in the 3rd field of the version
> identifier). That rationale seems valid to me and I'd rather not ignore it.
>
We recently had a discussion about this.  The overall idea is we want to
keep old releases available to Spack.   By default, Spack will "shepherd"
users to newer versions, which I believe does address your general
concerns.  But users who specify a desire to stick with old versions of
particular packages may also do so.

https://groups.google.com/forum/#!topic/spack/afI3-EbiglM

One important reason to keep old releases, even "bugfix" stuff, is to help
troubleshoot things.  Spack builds need to be *repeatable*.

> Can you not copy and save any release you value?
>
We have over 1000 packages and limited resources.  Upstream packages that
move their URLs around are a persistent problem because it means we have to
keep updating our packages; we usually try to find a way to download these
packages that doesn't change with time.  Otherwise, older verisons of Spack
will no longer work.  Suppose that we go onto a 3-month release cycle:
Spack 1.0 is released January 1, to be updated to Spack 1.1 April 1.  if
UDUNITS2 changes a URL January 5, then the latest official/tested release
of Spack will go almost 3 months without being able to install UDUNITS2.
Conclusion: if upstream URLs are changing all the time, there is really no
practical way we can ever make a Spack release that will work reliably.

We are considering having the Spack project maintain a cache of tarballs in
case of downed servers, but plan A should always be to download things
directly from the source.  This has many advantages: (1) less bandwidth at
LLNL, (2) does not usurp distribution control from upstream authors.  Many
authors rely on download statistics from end users, and us caching their
stuff breaks that.

-- Elizabeth

citibeth commented:
Summary: I think we should continue getting tarballs from where we've been getting them from; and remove 2.2.25 because it's not yet released.

michaelkuhn commented:
I have updated the PR to remove version 2.2.25 and added a short note for future versions.

@semmerson: One suggestions would be to simply not tag unreleased versions on GitHub and only add the tag once a version is final. This might help prevent some confusion for downstream distributions.

semmerson commented:
On Fri, Apr 14, 2017 at 1:08 PM, Todd Gamblin <notifications@github.com>
wrote:

> @semmerson <https://github.com/semmerson>: Question: Is a github label
> set in stone once its corresponding release appears on your FTP site?
>

Yes. To the best of my ability, the commit associated with a GitHub version
tag  shouldn't change once the corresponding release is official.

semmerson commented:
Elizabeth,

You *could* get previous version from the GitHub repository -- but that
would require more work on your part. Plus, building the package would be
different because a GitHub download isn't the same as a
source-distribution: the "m4/" directory must be made, autoreconf(1) must
be executed, etc.).

I don't suppose you'd want to do that.

semmerson commented:
Elizabeth,

Definitely remove 2.2.25.

Regards,
Steve Emmerson

citibeth commented:
Spack is good at running `autoreconf` when needed, so it doesn't make much
of a difference to us.  Having a stable URL is more important.


On Mon, Apr 17, 2017 at 12:53 PM, Steven Emmerson <notifications@github.com>
wrote:

> Elizabeth,
>
> You *could* get previous version from the GitHub repository -- but that
> would require more work on your part. Plus, building the package would be
> different because a GitHub download isn't the same as a
> source-distribution: the "m4/" directory must be made, autoreconf(1) must
> be executed, etc.).
>
> I don't suppose you'd want to do that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-294526926>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdz2M8iBro5gW5z6EatmdtrDc1Xsoks5rw5kIgaJpZM4M7uHs>
> .
>

semmerson commented:
Michael,

> @semmerson <https://github.com/semmerson>: One suggestions would be to
> simply not tag unreleased versions on GitHub and only add the tag once a
> version is final. This might help prevent some confusion for downstream
> distributions.
>
I have a continuous-deployment pipeline for the UDUNITS package. If a
pushed commit passes all the tests, then it automatically gets
published/released/deployed. This is easier if the commit is already tagged
coming-in.

I'll take a look at the pipeline. Something might be possible.

semmerson commented:
@michaelkuhn What triggers your process when I commit to GitHub? A push to the master branch or the addition of a version tag?

adamjstewart commented:
Whatever process in your build pipeline that creates a new numbered release on the releases page.

semmerson commented:
I interpret the comment by @adamjstewart to mean that the Spack process is triggered when a new release appears on the FTP site. This seems inconsistent with the solution proposed by @michaelkuhn about not tagging the commit until a release is made. If the processes isn't triggered until a release is made, then it shouldn't matter if the associated tag is moved before the release.

Please help me out here.

citibeth commented:
AFAIK, there is no Spack process for adding new releases to packages.  They
get added when someone wants them and submits a PR.

On Thu, May 11, 2017 at 11:25 AM, Steven Emmerson <notifications@github.com>
wrote:

> I interpret the comment by @adamjstewart <https://github.com/adamjstewart>
> to mean that the Spack process is triggered when a new release appears on
> the FTP site. This seems inconsistent with the solution proposed by
> @michaelkuhn <https://github.com/michaelkuhn> about not tagging the
> commit until a release is made. If the processes isn't triggered until a
> release is made, then it shouldn't matter if the associated tag is moved
> before the release.
>
> Please help me out here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-300824990>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd_3SFIE83edp2pUvkWk3nncgt1aSks5r4yhwgaJpZM4M7uHs>
> .
>

adamjstewart commented:
Sorry, to be more clear, I was talking about the releases page on GitHub. And yes, @citibeth is right. Basically, if someone checks the udunits GitHub releases page and sees that there is a new release out, they will try to add it to Spack. If this release isn't really a release and the tag keeps moving, we will continue to run into the problem we previously discussed. If you could refrain from tagging releases on GitHub until they are finalized so that the tag doesn't move, it would help a lot. If you could make old releases available on the FTP site, that would be even better.

citibeth commented:
Why not just tag the new release on GitHub as `1.4beta`?  When it's
finalized and you generate the tarball, change the tag to `1.4`.

On Thu, May 11, 2017 at 11:33 AM, Adam J. Stewart <notifications@github.com>
wrote:

> Sorry, to be more clear, I was talking about the releases page on GitHub.
> And yes, @citibeth <https://github.com/citibeth> is right. Basically, if
> someone checks the udunits GitHub releases page and sees that there is a
> new release out, they will try to add it to Spack. If this release isn't
> really a release and the tag keeps moving, we will continue to run into the
> problem we previously discussed. If you could refrain from tagging releases
> on GitHub until they are finalized so that the tag doesn't move, it would
> help a lot. If you could make old releases available on the FTP site, that
> would be even better.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3809#issuecomment-300827376>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd53uo4o9PazQhz_pJG_D-5SGgw2Sks5r4yougaJpZM4M7uHs>
> .
>

semmerson commented:
Ah! I see the problem. The GitHub "Releases" webpage is being checked rather than the official FTP site (ftp://ftp.unidata.ucar.edu/pub/udunits). The "releases" listed on the GitHub webpage aren't releases: they're merely placeholders (see [this SlashDot thread](http://stackoverflow.com/questions/28496319/github-a-tag-but-not-a-release)). Only the FTP site has releases.

Changing where the releases are is problematical.

adamjstewart commented:
We don't mind changing where we download releases from, but our main concern is a stable website to download from. The FTP site works great, until all of a sudden the version you've been downloading from disappears and you don't know why. The GitHub site is more stable as all older releases are still available.

citibeth commented:
@adamjstewart I think we are repeating ourselves, to little avail at this
point.

I suggest we resolve this issue by making sure we look on the FTP site to
ensure that a release is out of Beta, before adding it to the Spack
package.  We will continue to use a Git fetch method for all releases.  We
add comments to the Spack package explaining this state of affairs,
including a reference to this conversation.

Yes it's unusual; but what else can we do?

semmerson commented:
If it's any consolation, I've removed the deleting of old bugfix versions on the FTP site. The UDUNITS package is small enough that I can do this.

adamjstewart commented:
Thanks @semmerson, that's great! If you're willing to add 2.2.21 and 2.2.23 to the FTP site, I would be willing to change our Spack package to point to the FTP site instead of GitHub.

semmerson commented:
@adamjstewart Done.
